File: src/applications/standalone/kvs_client/kvs_client.py
#!/usr/bin/env python3
from derecho.cascade.client import ServiceClientAPI

class bcolors:
    OK = '\033[92m' #GREEN
    WARNING = '\033[93m' #YELLOW
    FAIL = '\033[91m' #RED
    RESET = '\033[0m' #RESET COLOR

if __name__ == '__main__':
    print("KVS Client Example in Python.")

    print("1) Load configuration and connecting to cascade service...")
    capi = ServiceClientAPI()
    print("- connected.")

    print("2) Create a folder, a.k.a object pool in the first VolatileCascadeStore subgroup.")
    res = capi.create_object_pool("/vcss_objects","VolatileCascadeStoreWithStringKey",0)
    if res:
        ver = res.get_result()
        print(bcolors.OK + f"folder '/vcss_objects' is created with version:{ver}" + bcolors.RESET)
    else:
        print(bcolors.FAIL + "Something went wrong, create_object_pool returns null." + bcolors.RESET)
        quit()

    print("3) List all folders a.k.a. object pools:")
    res = capi.list_object_pools()
    print(bcolors.OK + f"{res}" + bcolors.RESET)

    print("4) Put an object with key '/vcss_objects/obj_001'")
    res = capi.put('/vcss_objects/obj_001',bytes('value of /vcss_objects/obj_001','utf-8'),previous_version=ServiceClientAPI.CURRENT_VERSION,previous_version_by_key=ServiceClientAPI.CURRENT_VERSION)
    if res:
        ver = res.get_result()
        print(bcolors.OK + f"Put is successful with version {ver}." + bcolors.RESET)
    else:
        print(bcolors.FAIL + "Something went wrong, put returns null." + bcolors.RESET)
        quit()

    print("5) Get an object with key '/vcss_objects/obj_001'")
    res = capi.get('/vcss_objects/obj_001')
    if res:
        odict = res.get_result()
        print(bcolors.OK + f"Get is successful with details: {odict}" + bcolors.RESET)
    else:
        print(bcolors.FAIL + "Something went wrong, get returns null." + bcolors.RESET)
        quit()

File: src/applications/standalone/kvs_client/kvs_client.cpp
#include <cascade/service_client_api.hpp>
#include <iostream>
#include <string>
#include <fstream>
#include <typeindex>
#include <sys/prctl.h>

using namespace derecho::cascade;

#define PROC_NAME "kvs_client"

int main(int argc, char** argv) {
    std::cout << "KVS Client Example in C++." << std::endl;

    std::cout << "1) Load configuration and connecting to cascade service..." << std::endl;
    ServiceClientAPI& capi = ServiceClientAPI::get_service_client();
    std::cout << "- connected." << std::endl;

#define OBJECT_FOLDER   "/vcss_objects"
    std::cout << "2) Create a folder, a.k.a. object pool in the first VolatileCascadeStore subgroup." << std::endl;
    auto result_2 = capi.template create_object_pool<VolatileCascadeStoreWithStringKey>(OBJECT_FOLDER,0);
    for (auto& reply_future:result_2.get()) {
        auto reply = reply_future.second.get();
        std::cout << "node(" << reply_future.first << ") replied with version:" << std::get<0>(reply)
                  << ",ts_us:" << std::get<1>(reply)
                  << std::endl;
    }
    std::cout << "- " << OBJECT_FOLDER <<" folder is created." << std::endl;

    std::cout << "3) List all folders a.k.a. object pools:" << std::endl;
    for (auto& op: capi.list_object_pools(true)) {
        std::cout << "\t" << op << std::endl;
    }

#define OBJECT_KEY      OBJECT_FOLDER "/obj_001"
    std::cout << "4) Put an object with key '" << OBJECT_KEY << "'" << std::endl;
#define OBJECT_VALUE    "value of " OBJECT_KEY
    ObjectWithStringKey obj;
    obj.key = OBJECT_KEY;
    obj.previous_version = INVALID_VERSION;
    obj.previous_version_by_key = INVALID_VERSION;
    obj.blob = Blob(reinterpret_cast<const uint8_t*>(OBJECT_VALUE),std::strlen(OBJECT_VALUE));
    auto result_4 = capi.put(obj);
    for (auto& reply_future:result_4.get()) {
        auto reply = reply_future.second.get();
        std::cout << "node(" << reply_future.first << ") replied with version:" << std::get<0>(reply)
                  << ",ts_us:" << std::get<1>(reply)
                  << std::endl;
    }

    std::cout << "5) Get an object with key '" << OBJECT_KEY << "'" << std::endl;
    auto result_5 = capi.get(std::string(OBJECT_KEY),CURRENT_VERSION);
    for (auto& reply_future:result_5.get()) {
        auto reply = reply_future.second.get();
        std::cout << "node(" << reply_future.first << ") replied with value:" << reply << std::endl;
    }
    return 0;
}
